### PR TITLE
RMN Controller Improvements

### PR DESCRIPTION
**Pull request by dimkouv**
- Improved logging.
- In the Outcome phase exclude source chains that are not signed to prevent on-chain revert.
- Add comments explaining critical components.
- Trigger additional RMN requests instantly on the first error instead of waiting for timer expiration.
- Fix bug where duplicate lane updates were added in the same RMN report prior to verifying sigs.
- Cleaner less complicated logic.
- Handle edge case where some RMN node mix valid and invalid roots for different chains in their observation.
- Stop earlier after all RMN responses are received instead of waiting for timer expiration (on signature requests).

### CODE REVIEW

File: commit/merkleroot/outcome.go [hunk 1]
```diff
@@ -138,15 +139,34 @@ func buildReport(
...
-		lggr.Warn("RMNSignatures are nil")
-	} else {
+	if q.RMNSignatures != nil { // TODO: should never be nil, error after e2e RMN integration.
 		parsedSigs, err := rmn.NewECDSASigsFromPB(q.RMNSignatures.Signatures)
 		if err != nil {
-			lggr.Errorw("Failed to parse RMN signatures", "error", err)
-		} else {
-			sigs = parsedSigs
+			lggr.Errorw("Failed to parse RMN signatures returning an empty outcome", "error", err)
+			return Outcome{}
```
Comment (makramkd) [inline commit/merkleroot/outcome.go:146]: Should this situation return an error to the top-level plugin?

File: commit/merkleroot/outcome.go [hunk 2]
```diff
@@ -138,15 +139,34 @@ func buildReport(
...
 		}
+		sigs = parsedSigs
+
+		signedRoots := mapset.NewSet[cciptypes.MerkleRootChain]()
+		for _, laneUpdate := range q.RMNSignatures.LaneUpdates {
```
Comment (makramkd) [inline commit/merkleroot/outcome.go:151]: I'm not sure if we do these validations elsewhere, but one thing I can think of is do we validate that RMN nodes don't send overlapping intervals / different roots for same interval / etc. ?

File: commit/merkleroot/rmn/controller.go [hunk 3]
```diff
@@ -323,14 +347,53 @@ func (c *controller) listenForRmnObservationResponses(
...
+		sort.Slice(values, func(i, j int) bool { return values[i] < values[j] })
+		if values[len(values)-1] < minObservers {
+			return false
+		}
+	}
+
+	return true
+}
+
 // nolint:gocyclo // todo
```
Comment (makramkd) [inline commit/merkleroot/rmn/controller.go:384]: Didn't notice this `nolint` before, is there any way we can reduce cyclomatic complexity of this func?

File: commit/merkleroot/rmn/controller.go [hunk 4]
```diff
@@ -463,70 +625,22 @@ func (c *controller) getRmnReportSignatures(
...
-	return &ReportSignatures{
-		Signatures:  ecdsaSignatures,
-		LaneUpdates: fixedDestLaneUpdates,
-	}, nil
+	return requestIDs, signersRequested, nil
 }
 
-// reportSigWithNodeID is a helper struct to store the report signature and the node ID that provided it.
-type reportSigWithNodeID struct {
+// reportSigWithSignerAddress is a helper struct to store the report signature and the address of the SK that signed it.
```
Comment (makramkd) [inline commit/merkleroot/rmn/controller.go:642]:
```suggestion
// reportSigWithSignerAddress is a helper struct to store the report signature and the address of the private key that signed it.
```

File: commit/merkleroot/outcome.go [hunk 5]
```diff
@@ -138,15 +139,34 @@ func buildReport(
...
+		}
+
+		// Only report roots that are present in RMN signatures.
+		rootsToReport := make([]cciptypes.MerkleRootChain, 0)
+		for _, root := range roots {
+			if signedRoots.Contains(root) {
+				rootsToReport = append(rootsToReport, root)
+			}
+		}
+		roots = rootsToReport
```
Comment (makramkd) [inline commit/merkleroot/outcome.go:169]: Lets log the excluded roots, if any

File: commit/merkleroot/outcome.go (hunk 2 above)
Comment (dimkouv) [reply commit/merkleroot/outcome.go:151]: Yes we handle all this

File: commit/merkleroot/rmn/controller.go (hunk 3 above)
Comment (dimkouv) [reply commit/merkleroot/rmn/controller.go:384]: It's a bit hard actually, i'll try in a follow up.
I enable gocyclo linter recently.

File: commit/merkleroot/outcome.go (hunk 1 above)
Comment (dimkouv) [reply commit/merkleroot/outcome.go:146]: The way plugin is designed, it's the same thing, just that plugin would log the error.

**Review by makramkd — APPROVED** [review]

**Pull request merged**